### ISSUE TITLE
Allow partial application with keyword arguments

### ISSUE THREAD

**Issue by E1k3** [omni-us/jsonargparse#755]
## 🚀 Feature request
The current implementation of partial application for callables as described [here](https://jsonargparse.readthedocs.io/en/v4.40.2/#callable-type) only allows for positional arguments.
 Would it be possible to extend the partial application support to allow for keyword arguments?
 
### Motivation
After our discussion in https://github.com/omni-us/jsonargparse/issues/745 I started using partial application to instantiate objects after dependency injection, but I started noticing a few problems.

Since the arguments of callable types in code are unnamed, they don't document what each argument means semantically: 
In `dataloader: Callable[[Dataset, int], DataLoader]` the `int` is the `batch_size`, but thats not visible in the code.

For constructors / functions of library code, I regularly had to write wrappers to reorder arguments, so they begin with the arguments that aren't available at injection time.
Let's say, I want to provide the `dataset` and `shuffle` arguments of `DataLoader` after injection, I have to create something like this:
```py
class DataModule(LightningDataModule):
    def __init__(self, dataloader_factory: Callable[[Dataset, bool], torch.utils.data.DataLoader]):
        pass

class DataloaderWrapper(torch.utils.data.DataLoader):
    def __init__(self, dataset: Dataset, shuffle: bool, **kwargs):
        super().__init__(
            dataset=dataset,
            shuffle=shuffle,
            **kwargs
        )

# or
def dataloader_factory(dataset: Dataset, shuffle: bool, **kwargs):
    return torch.utils.data.DataLoader(
        dataset=dataset,
        shuffle=shuffle,
        **kwargs
    )
```

### Pitch
Would it be possible to annotate the callable with `typing.Protocol` as described in the [Python documentation](https://docs.python.org/3.13/library/typing.html#annotating-callable-objects) and have jsonargparse translate this to a partial application with keyword arguments?
I understand that `Protocol` cannot be used directly here since it is supported by jsonargparse to constrain types, but a wrapper should be possible, right?

```py
class DataloaderFactory(jsonargparse.CallableProtocol):
    def __call__(self, dataset: Dataset, shuffle: bool) -> DataLoader: ...

class DataModule(LightningDataModule):
    def __init__(self, dataloader_factory: DataloaderFactory):
        pass
```

This way, DataloaderFactory can be defined right next to the DataModule and the signature including argument names is easily visible as well as checkable by jsonargparse and any static type checker.

**Comment by mauvilsa** [comment]
Thank you for the proposal!

> Would it be possible to annotate the callable with `typing.Protocol` as described in the [Python documentation](https://docs.python.org/3.13/library/typing.html#annotating-callable-objects) and have jsonargparse translate this to a partial application with keyword arguments?

This would be a great addition. I seem to recall that I was looking into how to use callable protocols to extend the instance factories dependency injection support. It was clear that the unnamed positionals is a big limitation. But I guess back then I was most concerned about optimizers and schedulers, so most likely I left it for the future.

> but a wrapper should be possible, right?
> 
> ```py
> class DataloaderFactory(jsonargparse.CallableProtocol):
>     def __call__(self, dataset: Dataset, shuffle: bool) -> None: ...
> ```

No, I would not consider this. Type hints should make sense independent of jsonargparse.

> I understand that `Protocol` cannot be used directly here since it is supported by jsonargparse to constrain types

I am not sure why you say this. Right now I have only thought for a short while, but nothing comes to mind that would prevent using normal callable protocols for this.

One thing for sure is that it would be a requirement that the `__call__` signature have as return type the base class of the instances it creates. The `-> None:` from above would not be correct.

**Comment by E1k3** [comment]
I thought using standard protocols would somehow interfere with the protocol support currently built into jsonargparse, but you're right, I cannot think of a case where that would happen either.
Should be straight forward to detect cases where a protocol is expected, but a function / type is given with incomplete arguments.

Regarding the return type, you're right of course. I edited the example code.

EDIT: nevermind about the edgecase. I guess that could only happen with `Any` as return type of the callable or callable objects returning self.
